x/wasm/keeper/ante.go

```diff
@@ -0,0 +1,40 @@
+package keeper
+
+import (
+	"github.com/CosmWasm/wasmd/x/wasm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type CountTXDecorator struct {
+	storeKey sdk.StoreKey
+}
+
+func NewCountTXDecorator(storeKey sdk.StoreKey) *CountTXDecorator {
+	return &CountTXDecorator{storeKey: storeKey}
+}
+
+func (a CountTXDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	if simulate {
+		return next(types.WithTXCounter(ctx, 0), tx, simulate)
+	}
+	currentHeight := ctx.BlockHeight()
+	store := ctx.KVStore(a.storeKey)
+	var txCounter uint64 = 0
+	if bz := store.Get(types.TXCounterPrefix); bz != nil {
+		lastHeight, val := decodeHeightCounter(bz)
+		if currentHeight == lastHeight {
+			txCounter = val
+		}
+	}
+	store.Set(types.TXCounterPrefix, encodeHeightCounter(currentHeight, txCounter+1))
+
+	return next(types.WithTXCounter(ctx, txCounter), tx, simulate)
+}
+
+func encodeHeightCounter(height int64, counter uint64) []byte {
+	return append(sdk.Uint64ToBigEndian(uint64(height)), sdk.Uint64ToBigEndian(counter)...)
```

Comment:
I would use a `uint32` type for the counter now but otherwise leave it as it is. We can always refactor as needed. Depending on the case we another ante handler may be a better option to keep things separated. 🤷